### PR TITLE
Fix various memory leaks in unifyfsd

### PR DESCRIPTION
**Pull request by rgmiller**
This is a collection of commits that fix a variety of memory leaks in the daemon.
### Description

There were a variety of memory leaks found by running valgrind against the daemon.  Most of the leaks were minor (failing to release memory when the daemon exited), but a few did seem to leak memory continuously as the daemon ran.

### Motivation and Context

Resolves issue #660 

### How Has This Been Tested?

Testing was done by running unifyfsd within the context of valgrind on an Ubuntu workstation.  A few simple clients were run to read & write data and then the daemon was killed.  The resulting valgrind report was analyzed.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Testing (addition of new tests or update to current tests)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the UnifyFS code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted.

### CODE REVIEW

**Review by MichaelBrim — CHANGES_REQUESTED** [review]
For the `unifyfs_inode_tree_clear()` changes, use`unifyfs_inode_destroy(oldnode)` instead of freeing the allocations inside the inode directly.

File: server/src/unifyfs_server.c [hunk 1]
```diff
@@ -675,6 +678,11 @@ static int unifyfs_exit(void)
     LOGDBG("stopping rpc service");
     margo_server_rpc_finalize();
 
+    ret = ABT_mutex_free(&app_configs_abt_sync);
```
Comment (MichaelBrim) [inline server/src/unifyfs_server.c:681]: Not sure about this, but I think if `margo_finalize()` is being called as part of `margo_server_rpc_finalize()` above, and we're relying on Margo to init/fini Argobots, then we may not be able to safely call any ABT_xxx methods after that.

**Comment by rgmiller** [comment]
> For the `unifyfs_inode_tree_clear()` changes, use`unifyfs_inode_destroy(oldnode)` instead of freeing the allocations inside the inode directly.

That will require some changes in unifyfs_inode.c:  `unifyfs_inode_destroy()` is currently a static function.  If we want to call it from `unifyfs_inode_tree_clear()`, we'd have to change that.  I don't know if that will cause issues elsewhere or not.

**Comment by MichaelBrim** [comment]
@rgmiller if you are happy with the current code, please do a rebase of this versus current dev and squash down to one commit.

**Comment by rgmiller** [comment]
> @rgmiller if you are happy with the current code, please do a rebase of this versus current dev and squash down to one commit.

@MichaelBrim Done - sort of.  It seems to have squashed down to 4 commits.  The fskv_init thing has already been merged in and two others were merges of the latest dev branch changes back into the dev branch on my fork.  I'm guessing that's why they didn't get squashed.  Why the fskv_init commit is showing up at all still baffles me.

**Comment by rgmiller** [comment]
I resolved the conflict in unifyfs_inode.c.  Not sure about unifyfs_fid.c.

**Comment by MichaelBrim** [comment]
@rgmiller Because of the intermediate merges, it doesn't appear there is an easy way to get rgmiller:dev back in a state where there is a single commit versus LLNL:dev. My suggestion would be to start over with a new branch based on LLNL:dev and re-do the necessary changes in that branch, and then submit a new PR.

**Comment by rgmiller** [comment]
What about just squashing the commits during the merge?  GitHub has an option to do that, doesn't it?

**Comment by MichaelBrim** [comment]
Currently this has a bunch of "changes" unrelated to your memory leak fixes, which makes it hard to track where the code came from and why it changed. In general though, we try to avoid having merge commits in the history. I can help clean this up if you want me to.

**Comment by rgmiller** [comment]
OK, I've created a new branch (in my forked repo) with just the memory leak fixes.  I'll close this pull request and create a new one using the new branch.

**Pull request closed without merging**